File: funds/scripts/classify_funds.js
/**
 * Fund Classification Generator
 * DC형 70% 위험자산 한도 자동 검증을 위한 펀드 분류 데이터 생성
 * 
 * 위험자산: 주식형, 해외주식형, 주식혼합형, 해외주식혼합형, 채권혼합형, 해외채권혼합형
 * 안전자산: 채권형, 해외채권형, 기타 (MMF, 예금, 골드, TDF 등)
 */

const fs = require('fs');
const path = require('path');

// 위험자산 카테고리
const RISK_ASSET_CATEGORIES = [
  '주식형',
  '해외주식형',
  '주식혼합형',
  '해외주식혼합형',
  '채권혼합형',
  '해외채권혼합형'
];

// 안전자산 카테고리
const SAFE_ASSET_CATEGORIES = [
  '채권형',
  '해외채권형',
  '기타'
];

// 펀드명으로 카테고리 추정 (generate_md.js 로직 기반)
function getFundCategory(fundName) {
  const name = fundName;
  
  // 1. 해외 여부 확인
  const isOverseas = /해외|글로벌|미국|차이나|인디아|베트남|일본|유럽|아시아|이머징|Global|USA|China|India|Japan|Europe|Asia|Emerging/i.test(name);
  
  // 2. 자산 유형 확인
  const hasBondMixed = /채권혼합|채권 혼합/.test(name);
  const hasStockMixed = /주식혼합|주식 혼합/.test(name);
  const hasBond = /채권|크레딧|국공채|인컴|하이일드|단기채/.test(name) && !hasBondMixed;
  const hasStock = /주식/.test(name) && !hasStockMixed && !hasBondMixed;
  
  // 3. 특수 유형 확인
  const isSpecial = /TDF|MMF|EMP|리츠|골드|부동산|원자재|금-재간접|TIF/.test(name);
  
  // 분류 결정
  if (isSpecial) {
    return '기타';
  }
  
  if (isOverseas) {
    if (hasBondMixed) return '해외채권혼합형';
    if (hasStockMixed) return '해외주식혼합형';
    if (hasBond) return '해외채권형';
    if (hasStock) return '해외주식형';
    // 기본: 해외주식형으로 추정 (대부분 해외펀드는 주식형)
    return '해외주식형';
  } else {
    if (hasBondMixed) return '채권혼합형';
    if (hasStockMixed) return '주식혼합형';
    if (hasBond) return '채권형';
    if (hasStock) return '주식형';
    return '기타';
  }
}

// 테마 추출
function extractTheme(fundName) {
  const themes = {
    'semiconductor': /반도체|필라델피아/i.test(fundName),
    'ai': /AI|인공지능|Chat AI|메타버스/i.test(fundName),
    'robot': /로봇|휴머노이드|로보틱스|로보테크/i.test(fundName),
    'healthcare': /헬스케어|바이오|헬스사이언스/i.test(fundName),
    'dividend': /배당|고배당|인컴/i.test(fundName),
    'value': /가치|밸류/i.test(fundName),
    'growth': /성장|그로스/i.test(fundName),
    'index': /인덱스|KOSPI|S&P|나스닥|KRX/i.test(fundName),
    'ev': /전기차|자율주행|모빌리티/i.test(fundName),
    'gold': /골드|금/i.test(fundName),
    'energy': /에너지|그린|클린테크/i.test(fundName),
    'reits': /리츠|부동산/i.test(fundName),
    'space': /우주항공/i.test(fundName)
  };
  
  const matched = Object.entries(themes)
    .filter(([_, match]) => match)
    .map(([theme]) => theme);
  
  return matched.length > 0 ? matched : ['general'];
}

// 지역 추출
function extractRegion(fundName) {
  if (/미국|USA|S&P|나스닥|Nasdaq/i.test(fundName)) return 'us';
  if (/차이나|중국|China/i.test(fundName)) return 'china';
  if (/인디아|인도|India/i.test(fundName)) return 'india';
  if (/일본|Japan|재팬/i.test(fundName)) return 'japan';
  if (/유럽|Europe|유로/i.test(fundName)) return 'europe';
  if (/베트남|Vietnam/i.test(fundName)) return 'vietnam';
  if (/아세안|ASEAN/i.test(fundName)) return 'asean';
  if (/아시아|Asia/i.test(fundName)) return 'asia';
  if (/브라질|Brazil/i.test(fundName)) return 'brazil';
  if (/이머징|Emerging/i.test(fundName)) return 'emerging';
  if (/글로벌|Global|월드|World/i.test(fundName)) return 'global';
  if (/코리아|한국|KOSPI|KRX|코스닥/i.test(fundName)) return 'korea';
  return 'korea'; // 기본값
}

// 환헤지 여부
function isHedged(fundName) {
  if (/\(H\)|\[H\]|H[)]|\(환헤지\)/i.test(fundName)) return true;
  if (/\(UH\)|\[UH\]|UH[)]|\(환노출\)/i.test(fundName)) return false;
  return null; // 명시되지 않음
}

// 메인 실행
function main() {
  // fund_data.json 읽기
  const fundDataPath = path.join(__dirname, '..', 'fund_data.json');
  const funds = JSON.parse(fs.readFileSync(fundDataPath, 'utf8'));
  
  // 분류 데이터 생성
  const classification = {};
  
  for (const fund of funds) {
    const category = getFundCategory(fund.name);
    const isRiskAsset = RISK_ASSET_CATEGORIES.includes(category);
    
    classification[fund.name] = {
      category: category,
      riskAsset: isRiskAsset,
      assetClass: category.includes('채권') ? 'bond' : 
                  category.includes('혼합') ? 'mixed' : 
                  category.includes('주식') ? 'equity' : 'other',
      region: extractRegion(fund.name),
      themes: extractTheme(fund.name),
      hedged: isHedged(fund.name),
      riskLevel: fund.riskLevel,
      source: 'fund_data.json + keyword classification',
      generatedAt: new Date().toISOString().split('T')[0]
    };
  }
  
  // 통계 출력
  const stats = {
    total: funds.length,
    riskAssets: Object.values(classification).filter(c => c.riskAsset).length,
    safeAssets: Object.values(classification).filter(c => !c.riskAsset).length,
    byCategory: {}
  };
  
  for (const c of Object.values(classification)) {
    stats.byCategory[c.category] = (stats.byCategory[c.category] || 0) + 1;
  }
  
  console.log('=== Fund Classification Stats ===');
  console.log(`Total funds: ${stats.total}`);
  console.log(`Risk assets: ${stats.riskAssets} (${(stats.riskAssets/stats.total*100).toFixed(1)}%)`);
  console.log(`Safe assets: ${stats.safeAssets} (${(stats.safeAssets/stats.total*100).toFixed(1)}%)`);
  console.log('\nBy category:');
  for (const [cat, count] of Object.entries(stats.byCategory).sort((a,b) => b[1]-a[1])) {
    const riskLabel = RISK_ASSET_CATEGORIES.includes(cat) ? '[RISK]' : '[SAFE]';
    console.log(`  ${cat}: ${count} ${riskLabel}`);
  }
  
  // 파일 저장
  const outputPath = path.join(__dirname, '..', 'fund_classification.json');
  fs.writeFileSync(outputPath, JSON.stringify(classification, null, 2), 'utf8');
  console.log(`\nSaved to: ${outputPath}`);
  
  return classification;
}

main();

File: scripts/classify_funds.py
#!/usr/bin/env python3
"""
Enhanced Fund Classification Script
Classifies funds into risk/safe assets with improved keyword matching.
"""

import json
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Tuple, List


def classify_fund(name: str, risk_level: int) -> Tuple[str, bool, str, str, List[str]]:
    """
    Classify a fund based on its name and risk level.
    Returns: (category, is_risk_asset, asset_class, region, themes)
    """
    name_lower = name.lower()

    # Default classification based on risk level
    is_risk_asset = risk_level <= 3  # 1-3등급 = 위험자산
    asset_class = "mixed"
    region = "domestic"
    themes = []
    category = "기타"

    # Priority 1: Specific thematic keywords (highest priority)
    thematic_keywords = {
        # Technology & Semiconductors
        "반도체": ("주식형", True, "equity", "global", ["semiconductor"]),
        "HBM": ("해외주식형", True, "equity", "global", ["semiconductor", "ai"]),
        "SK하이닉스": ("주식형", True, "equity", "domestic", ["semiconductor"]),
        "밸류체인": ("주식형", True, "equity", "domestic", ["technology"]),
        # Defense & Nuclear
        "방산": ("주식형", True, "equity", "domestic", ["defense"]),
        "우주": ("주식형", True, "equity", "domestic", ["defense", "space"]),
        "원자력": ("주식형", True, "equity", "domestic", ["nuclear"]),
        # Energy & Infrastructure
        "전력": ("주식형", True, "equity", "domestic", ["energy"]),
        "태양광": ("주식형", True, "equity", "domestic", ["renewable"]),
        "ESS": ("주식형", True, "equity", "domestic", ["energy"]),
        "신재생": ("주식형", True, "equity", "domestic", ["renewable"]),
        "수소": ("주식형", True, "equity", "domestic", ["hydrogen"]),
        # Other themes
        "금": ("해외주식형", True, "equity", "global", ["gold", "commodity"]),
        "골드": ("해외주식형", True, "equity", "global", ["gold", "commodity"]),
        "로봇": ("주식형", True, "equity", "domestic", ["robotics"]),
        "AI": ("해외주식형", True, "equity", "global", ["ai"]),
    }

    for keyword, (cat, risk, asset, reg, theme) in thematic_keywords.items():
        if keyword in name:
            return cat, risk, asset, reg, theme

    # Priority 2: Broad equity classification
    equity_keywords = {
        # Domestic equity indicators
        "KODEX": ("주식형", True, "equity", "domestic", []),
        "TIGER": ("주식형", True, "equity", "domestic", []),
        "ARIRANG": ("주식형", True, "equity", "domestic", []),
        "KINDEX": ("주식형", True, "equity", "domestic", []),
        "KBSTAR": ("주식형", True, "equity", "domestic", []),
        "RISE": ("주식형", True, "equity", "domestic", []),
        "HANARO": ("주식형", True, "equity", "domestic", []),
        "PLUS": ("주식형", True, "equity", "domestic", []),
        "UNICORN": ("주식형", True, "equity", "domestic", []),
        "SOL": ("주식형", True, "equity", "domestic", []),
        # Domestic indicators in name
        "코리아": ("주식형", True, "equity", "domestic", []),
        "대한민국": ("주식형", True, "equity", "domestic", []),
        "한국": ("주식형", True, "equity", "domestic", []),
        # Index-based
        "KOSPI": ("주식형", True, "equity", "domestic", []),
        "KOSDAQ": ("주식형", True, "equity", "domestic", []),
        "TOP": ("주식형", True, "equity", "domestic", []),
        # Global equity
        "글로벌": ("해외주식형", True, "equity", "global", []),
        "월드": ("해외주식형", True, "equity", "global", []),
        "WORLD": ("해외주식형", True, "equity", "global", []),
        "GLOBAL": ("해외주식형", True, "equity", "global", []),
        # Specific regions
        "미국": ("해외주식형", True, "equity", "global", ["us"]),
        "중국": ("해외주식형", True, "equity", "global", ["china"]),
        "일본": ("해외주식형", True, "equity", "global", ["japan"]),
        "유럽": ("해외주식형", True, "equity", "global", ["europe"]),
        "신흥국": ("해외주식형", True, "equity", "global", ["emerging"]),
        "S&P": ("해외주식형", True, "equity", "global", ["us"]),
        "NASDAQ": ("해외주식형", True, "equity", "global", ["us", "tech"]),
    }

    for keyword, (cat, risk, asset, reg, theme) in equity_keywords.items():
        if keyword in name:
            return cat, risk, asset, reg, theme

    # Priority 2.5: Bond-mixed assets (RISK assets per DC regulation)
    # 과학기술공제회 DC형 규정: 채권혼합형/해외채권혼합형은 위험자산으로 분류
    # 참조: "주식형>주식혼합형>채권혼합형 순서로 위험자산 비중이 높음"
    bond_mixed_keywords = {
        "채권혼합": ("채권혼합형", True, "mixed", "domestic", []),
        "해외채권혼합": ("해외채권혼합형", True, "mixed", "global", []),
    }

    for keyword, (cat, risk, asset, reg, theme) in bond_mixed_keywords.items():
        if keyword in name:
            return cat, risk, asset, reg, theme

    # Priority 3: Safe assets (bonds, MMF)
    safe_keywords = {
        "채권": ("채권형", False, "bond", "domestic", []),
        "국공채": ("채권형", False, "bond", "domestic", ["government"]),
        "회사채": ("채권형", False, "bond", "domestic", ["corporate"]),
        "크레딧": ("채권형", False, "bond", "domestic", ["credit"]),
        "MMF": ("MMF", False, "money_market", "domestic", []),
        "머니마켓": ("MMF", False, "money_market", "domestic", []),
        "단기금융": ("MMF", False, "money_market", "domestic", []),
    }

    for keyword, (cat, risk, asset, reg, theme) in safe_keywords.items():
        if keyword in name:
            return cat, risk, asset, reg, theme

    # Priority 4: Generic patterns
    if "증권상장지수투자신탁" in name and "[주식]" in name:
        # ETF with [주식] tag - likely equity ETF
        if "해외" in name or "Global" in name or "World" in name:
            return "해외주식형", True, "equity", "global", []
        else:
            return "주식형", True, "equity", "domestic", []

    if "증권투자신탁[주식]" in name or "증권자투자신탁(주식" in name:
        # Regular equity fund
        if "해외" in name or "Global" in name or "World" in name:
            return "해외주식형", True, "equity", "global", []
        else:
            return "주식형", True, "equity", "domestic", []

    # Priority 5: Fallback based on risk level
    if risk_level <= 3:
        # High risk, likely equity
        category = "주식형"
        is_risk_asset = True
        asset_class = "equity"
    elif risk_level >= 5:
        # Low risk, likely bond or conservative
        category = "채권형"
        is_risk_asset = False
        asset_class = "bond"
    else:
        # Medium risk, mixed
        category = "혼합형"
        is_risk_asset = True
        asset_class = "mixed"

    return category, is_risk_asset, asset_class, region, themes


def generate_classification(fund_data_path: str, output_path: str) -> Dict[str, Any]:
    """
    Generate enhanced fund classification file.
    """
    with open(fund_data_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    funds = data.get("funds", [])
    classifications = {}

    for fund in funds:
        name = fund["name"]
        risk_level = fund["riskLevel"]

        category, is_risk_asset, asset_class, region, themes = classify_fund(
            name, risk_level
        )

        # Check if hedged
        hedged = (
            "헤지" in name or "Hedged" in name.lower() or "(H)" in name or "[H]" in name
        )

        classifications[name] = {
            "category": category,
            "riskAsset": is_risk_asset,
            "assetClass": asset_class,
            "region": region,
            "themes": themes,
            "hedged": hedged,
            "riskLevel": risk_level,
            "source": "enhanced keyword classification v2",
            "generatedAt": datetime.now().strftime("%Y-%m-%d"),
        }

    # Write classification file
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(classifications, f, ensure_ascii=False, indent=2)

    # Calculate statistics
    risk_count = sum(1 for c in classifications.values() if c["riskAsset"])
    safe_count = len(classifications) - risk_count

    category_stats = {}
    for c in classifications.values():
        cat = c["category"]
        category_stats[cat] = category_stats.get(cat, 0) + 1

    return {
        "total": len(classifications),
        "risk_assets": risk_count,
        "safe_assets": safe_count,
        "categories": category_stats,
        "output_path": output_path,
    }


def main():
    import argparse

    parser = argparse.ArgumentParser(
        description="Enhanced fund classification with improved keyword matching"
    )
    parser.add_argument("--fund-data", required=True, help="Path to fund_data.json")
    parser.add_argument(
        "--output",
        help="Output path for classification JSON (default: same dir as fund_data.json)",
    )

    args = parser.parse_args()

    fund_data_path = Path(args.fund_data)
    if not fund_data_path.exists():
        print(f"Error: {fund_data_path} not found", file=sys.stderr)
        sys.exit(1)

    if args.output:
        output_path = Path(args.output)
    else:
        output_path = fund_data_path.parent / "fund_classification.json"

    print(f"Input: {fund_data_path}")
    print(f"Output: {output_path}")
    print("-" * 60)

    try:
        stats = generate_classification(str(fund_data_path), str(output_path))

        print("\nClassification completed successfully!")
        print(f"Total funds: {stats['total']}")
        print(
            f"Risk assets: {stats['risk_assets']} ({stats['risk_assets'] / stats['total'] * 100:.1f}%)"
        )
        print(
            f"Safe assets: {stats['safe_assets']} ({stats['safe_assets'] / stats['total'] * 100:.1f}%)"
        )
        print(f"\nCategory breakdown:")
        for cat, count in sorted(stats["categories"].items(), key=lambda x: -x[1]):
            pct = count / stats["total"] * 100
            print(f"  {cat}: {count} ({pct:.1f}%)")
        print(f"\nOutput: {stats['output_path']}")

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback

        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
